### PR TITLE
added STYLE.md

### PR DESCRIPTION
**Pull request by Rugvip**
This is a suggestion for a set of style and design guidelines. It starts out slim with some simple TS guidelines and then gets more opinionated :grin:

Especially the "API Design" section may contain some controversial bits @backstage/maintainers, although I hope nothing that's too surprising.

The documentation guidelines are taken from #7162, so that we're able to eventually close that issue

### CODE REVIEW

File: STYLE.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+Our TypeScript style is inspired by the [style guidelines](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines) of the TypeScript implementation itself.
+
+## Naming
+
+1. Use PascalCase for type names.
+1. Do not use `I` as a prefix for interface names.
+1. Use PascalCase for `enum` values.
+1. Use camelCase for function names.
+1. Use camelCase for property names and local variables.
+1. Do not use `_` as a prefix for private properties.
```
Comment (benjdlambert) [inline STYLE.md:16]: Does use `#` work as a good recommendation here?

**Review by benjdlambert — APPROVED** [review]

File: STYLE.md (hunk 1 above)
Comment (Rugvip) [reply STYLE.md:16]: Not at the moment, not supported by storybook yet ._.

**Review by jhaals — APPROVED** [review]
Love it!

File: STYLE.md (hunk 1 above)
Comment (benjdlambert) [reply STYLE.md:16]: Ah ok - I thought I saw a PR fixing that, but maybe it was just removing them :joy:

File: STYLE.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+Instead prefer to use a single parameter variable and then move the destructuring into the function body instead, which will look much nicer:
+
+![image](https://user-images.githubusercontent.com/4984472/133120542-d648de43-5495-43a8-b532-f5e6f6ab736e.png)
+
+Also be sure to check that the type used by the parameter is exported, as it otherwise won't be discoverable through documentation.
+
+## Use `@remarks` to split long descriptions
+
+The API reference has an index of exported symbols for each package, which uses a short description, while clicking through to the page for a symbol shows the full description. By default all descriptions are considered "short", and you have manually add a divider where the description should be cut off using the `@remarks` tag.
```
Comment (freben) [inline STYLE.md:173]:
```suggestion
The API reference has an index of exported symbols for each package, which uses a short description, while clicking through to the page for a symbol shows the full description. By default all descriptions are considered "short", and you have to manually add a divider where the description should be cut off using the `@remarks` tag.
```

File: STYLE.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+ * @param pluginId - The plugin that the cache manager should be created for. Plugin names should be unique.
+ */
+forPlugin(pluginId: string): PluginCacheManager {
+  ...
+}
+```
+
+## Fill in missing references using `{@link ...}`
+
+Not all types are detected and referenced on the documentation page. Most notably variables, and therefore ApiRefs, will not have clickable links to other symbols that they reference. We instead fill in this missing information using inline `{@link ...}` tags in the description:
```
Comment (freben) [inline STYLE.md:208]:
```suggestion
Not all types are detected and referenced on the documentation page. Most notably variables, and therefore `ApiRef`s, will not have clickable links to other symbols that they reference. We instead fill in this missing information using inline `{@link ...}` tags in the description:
```

**Review by freben — APPROVED** [review]
Awesome!

File: STYLE.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+## Syntax and Types
+
+1. Use `undefined`. Do not use `null`.
+1. Prefer `for..of` over `.forEach`.
+1. Do not introduce new types/values to the global namespace.
+
+## File and Export Structure
+
+1. Shared types should be defined in types.ts.
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply STYLE.md:27]: Tbh this is copied straight from https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines and I didn't give more thought to it than the fact that we're doing it in some form atm. I think the most important part is to put shared types in specifically `types.ts` and not use some other filename like `common.ts`, then what counts as a shared type is a slightly different discussion. Basically I'm thinking that this wording is enough as going into more detail might actually confuse things, but we could refine it over time if needed

**Pull request merged**